### DISCUSSION TITLE
Is there anyway to force updating GUI during the code excute in UI thread

### DISCUSSION THREAD

**Discussion by ruihuasun** [microsoft/microsoft-ui-xaml#8030] [Q&A]
I want to progammly select a new created item in treeview, however, the TreeViewItem/TreeViewNode of that item will not been built till the GUI update, also will not been built till the rootnode expand (if the rootnode is in collapsed state, but itemsource change).

The tree structure is:
rootnode
--node1
----subnode1
----subnode2
--node2

If I do something like that
```
rootnodeitem.NotifyPropertyChanged("ItemSouceBindData");      // It will update rootnode children.
treeview.Expand(rootnode);      // Try to make rootnode expand to make new TreeViewItem load
foreach(var node in rootnode.Children) treeview.Expand(node);
treeview.SelectedNode = treeview.NodefromContainer(treeview.ContainerfromItem(subnode1item));
```

Line3 and line4 are all useless, for after the `treeview.Expand(rootnode);`, node1.Children.Count=0, So `treeview.Expand(node)` will do nothing and `treeview.ContainerfromItem(subnode1item)` or `treeview.NodefromContainer(...)` will all return null.

So I wonder if I can do something like that
```
rootnodeitem.NotifyPropertyChanged("ItemSouceBindData");      // It will update rootnode children.
FoceRefreshGUI();
treeview.Expand(rootnode);      // Try to make rootnode expand to make new TreeViewItem load
FoceRefreshGUI();
foreach(var node in rootnode.Children) treeview.Expand(node);
FoceRefreshGUI();
treeview.SelectedNode = treeview.NodefromContainer(treeview.ContainerfromItem(subnode1item));
```

I try lots of things like use `await Task.Run(async () => { await Task.Delay(10000); });`, but it is hard to control the delay time and also it will not gurantee the GUI finish update during the delay time.

So I wonder is there anyway to force updating GUI during the code excute in UI thread?
Or Is there anyway to force treeview refresh to build all its new chilnodes?

Thank you very much if you have any idea

**Comment by FabricioGonzalez** [comment]
I think, the controls have a method called Update.
I never used it, so I'm not sure if it solves your problem

**Comment by marcelwgn** [comment]
There are some ways you can achieve this. The easiest way would be to call [`UpdateLayout`](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.uielement.updatelayout?view=windows-app-sdk-1.2). Alternatively, you could call `InvalidateArrange` and `InvalidateMeasure` on the parent control to force a rerendering though I'm not sure if that would happen immediately or during the next layout pass.
